source/common/grpc/async_client_manager_impl.h

```diff
@@ -50,18 +54,64 @@ class AsyncClientManagerImpl : public AsyncClientManager {
   AsyncClientManagerImpl(Upstream::ClusterManager& cm, ThreadLocal::Instance& tls,
                          TimeSource& time_source, Api::Api& api, const StatNames& stat_names);
 
+  RawAsyncClientSharedPtr
+  getOrCreateRawAsyncClient(const envoy::config::core::v3::GrpcService& config, Stats::Scope& scope,
+                            bool skip_cluster_check) override {
+    RawAsyncClientSharedPtr client;
+    client = raw_async_client_cache_->getCache(config);
+    if (client != nullptr) {
+      return client;
+    }
+    client = factoryForGrpcService(config, scope, skip_cluster_check)->create();
+    raw_async_client_cache_->setCache(config, client);
+    return client;
+  }
   // Grpc::AsyncClientManager
   AsyncClientFactoryPtr factoryForGrpcService(const envoy::config::core::v3::GrpcService& config,
                                               Stats::Scope& scope,
                                               bool skip_cluster_check) override;
 
 private:
+  class RawAsyncClientCache : public ThreadLocal::ThreadLocalObject {
+  public:
+    void setCache(const envoy::config::core::v3::GrpcService& config,
+                  const RawAsyncClientSharedPtr& client) {
+      std::uint64_t key = hashByType(config);
+      cache_[key] = client;
+    }
+    RawAsyncClientSharedPtr getCache(const envoy::config::core::v3::GrpcService& config) const {
+      std::uint64_t key = hashByType(config);
+      auto it = cache_.find(key);
+      if (it == cache_.end()) {
+        return nullptr;
+      }
+      return it->second;
+    }
+
+  private:
+    absl::flat_hash_map<uint64_t, RawAsyncClientSharedPtr> cache_;
+    uint64_t hashByType(const envoy::config::core::v3::GrpcService& config) const {
+      std::uint64_t key;
+      switch (config.target_specifier_case()) {
+      case envoy::config::core::v3::GrpcService::TargetSpecifierCase::kEnvoyGrpc:
+        key = MessageUtil::hash(config.envoy_grpc());
+        break;
+      case envoy::config::core::v3::GrpcService::TargetSpecifierCase::kGoogleGrpc:
+        key = MessageUtil::hash(config.google_grpc());
+        break;
+      default:
```

Comment:
can you drop the `default:` and explicitly handle all cases if you are not doing so already? If you drop the default, then the compiler will tell you if you are missing the handling of cases, so if someone adds a new target_specifier_case they'll know to come here.

Reply:
That's a good point.